dht/dht.go

```diff
@@ -166,6 +171,23 @@ func (dht *distributedHashTable) Addrs(n int) []wire.Address {
 			break
 		}
 	}
+
+	// Sort addresses in order of their XOR distance from our own address.
+	sort.Slice(addrs, func(i, j int) bool {
+		fstSignatory, err := addrs[i].Signatory()
+		if err != nil {
+			// This is a defensive check although it should never occur as it
+			// should not be posssible to insert an address into the DHT with an
+			// invalid signatory.
+			return false
+		}
+		sndSignatory, err := addrs[i].Signatory()
+		if err != nil {
+			return false
+		}
+		return dht.isCloser(fstSignatory, sndSignatory)
+	})
```

Comment:
This is a very expensive operation, and should be cached. That is, there should probably be two in-memory ways to access `addrs`: `addrs map[id.Signatory]wire.Address` and `addrsSorted []wire.Address` which are both updated side-by-side when addresses are added/removed. This prevents needing to do lots of hashing (happens when calling `addr.Signatory()` and sorting whenever calling `Addrs`, but also allows *O(1)* lookup of an address based on the signatory.

Reply:
As discussed, this only occurs when explicitly fetching `Addrs` as opposed to using subnets, but regardless has been moved to the insertion/deletion stage.